src/chunk.c

```diff
@@ -3376,47 +3367,107 @@ list_return_srf(FunctionCallInfo fcinfo)
-				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-				 errmsg("schema required when dropping chunks on distributed hypertables")));
+	List *fk_relids = NIL;
+	ListCell *lf;
+	List *cachedfkeys = NIL;
+	Relation table_rel = table_open(table_relid, AccessShareLock);
 
-	/* Make sure the schema is set in case it was not given on the command line */
-	FC_ARG(fcinfo, 2) = NameGetDatum(schema_name);
-	FC_NULL(fcinfo, 2) = false;
+	/* this list is from the relcache and can disappear with a cache flush, so
+	 * no further catalog access till we save the fk relids */
+	cachedfkeys = RelationGetFKeyList(table_rel);
+	foreach (lf, cachedfkeys)
+	{
+		ForeignKeyCacheInfo *cachedfk = (ForeignKeyCacheInfo *) lfirst(lf);
+
+		/* conrelid should always be that of the table we're considering */
+		Assert(cachedfk->conrelid == RelationGetRelid(table_rel));
+		fk_relids = lappend_oid(fk_relids, cachedfk->confrelid);
+	}
+	table_close(table_rel, AccessShareLock);
 
-	ts_cm_functions->func_call_on_data_nodes(fcinfo, data_node_oids);
+	/* We have a FK between hypertable H and PAR. Hypertable H has number of
+	 * chunks C1, C2, etc. When we execute "drop table C", PG acquires locks
+	 * on C and PAR. If we have a query as "select * from hypertable", this
+	 * acquires a lock on C and PAR as well. But the order of the locks is not
+	 * the same and results in deadlocks. - github issue 865 We hope to
+	 * alleviate the problem by acquiring a lock on PAR before executing the
+	 * drop table stmt. This is not fool-proof as we could have multiple
+	 * fkrelids and the order of lock acquisition for these could differ as
+	 * well. Do not unlock - let the transaction semantics take care of it. */
+	foreach (lf, fk_relids)
+		LockRelationOid(lfirst_oid(lf), AccessExclusiveLock);
+}
+
+/*
+ * Find either the hypertable or the materialized hypertable, if the relid is
+ * a continuous aggregate, for the relid.
+ *
+ * Will error if relid is not a hypertable or view (or cannot be found).
+ *
+ * If `allows_materialization` is false, it will also error if the hypertable
+ * found is a materialized hypertable.
+ */
+static Hypertable *
+find_hypertable_for_drop_by_relid(Cache *hcache, Oid relid, bool allows_materialization)
+{
+	Hypertable *ht = ts_hypertable_cache_get_entry(hcache, relid, CACHE_FLAG_NONE);
+	if (ht)
+	{
+		const ContinuousAggHypertableStatus status = ts_continuous_agg_hypertable_status(ht->fd.id);
+		if (!allows_materialization && (status & HypertableIsMaterialization))
+			elog(ERROR, "cannot drop chunks on a continuous aggregate materialization table");
+	}
+	else
+	{
+		/* Using the name to look up the continuous aggregate, but we should
+		 * refactor the code to use the relid directly. */
+		const char *const schema_name = get_namespace_name(get_rel_namespace(relid));
+		const char *const table_name = get_rel_name(relid);
+		ContinuousAgg *const cagg = ts_continuous_agg_find_userview_name(schema_name, table_name);
+		Assert(schema_name && table_name);
```

Comment:
Variable declaration section is usually separated by new line from assertions and body:
```suggestion
		ContinuousAgg *const cagg = ts_continuous_agg_find_userview_name(schema_name, table_name);
		
		Assert(schema_name && table_name);
```